lib/saml/url_service.rb

```diff
@@ -190,14 +189,23 @@ def add_query(url, params)
       end
     end
 
-    def create_redirect_application(uuid)
-      LoginRedirectApplication.create(
-        uuid: uuid, redirect_application: @redirect_application
+    # Initialize a new SAMLRequestTracker, if a valid previous SAML UUID is
+    # given, copy over the redirect and created_at timestamp.  This is useful
+    # for a user that has to go through the upleveling process.
+    def initialize_tracker(params, previous_saml_uuid: nil)
+      previous = previous_saml_uuid && SAMLRequestTracker.find(previous_saml_uuid)
+      redirect = previous&.payload_attr(:redirect) || Settings.ssoe.redirects[params[:application]]
+      # if created_at is set to nil (meaning no previous tracker to use), it
+      # will be initialized to the current time when it is saved
+      @tracker = SAMLRequestTracker.new(
```

Comment:
Instead of setting an instance variable in the method, it could return the `SAMLRequestTracker`.  I think it would add some clarity to the initializer what this is doing by calling `@tracker = initialize_tracker(params, previous_saml_uuid: previous_saml_uuid)`.  If this method was named something like `build_saml_request_tracker_from_params` future readers wouldn't need to dig into the implementation unless they specifically cared.  This is more style, I suppose.

Reply:
originally I did just that, but then I started reading other parts of the code and noticed that `initialize_query_params` (called from the constructor) was setting the instance variable rather than returning it.  So for consitency between the two methods, I stoped returning the object.  Do you think it be better to convert both methods to returning values rather than setting instance variables?

Reply:
Yeah, I think so, but I didn't want to comment on code you didn't write.  If you're up for it, I endorse it!

Reply:
I ended up just moving the `initialize_query_params` logic to the `initialize` method because it was really simple and the function name was slightly misleading.  Let me know what you think of the changes https://github.com/department-of-veterans-affairs/vets-api/pull/4120/commits/8cd6f4126b0e0262636da084c6ffc588fcdbe447